File: examples/go/clean-api-go/application/queries/get_stock_movements.go
package queries

import (
	"context"

	bberrors "github.com/FreakJazz/backbone/backbone-go/errors"
	"github.com/FreakJazz/backbone/backbone-go/infrastructure/logging"
	bbex "github.com/FreakJazz/backbone/backbone-go/interfaces/responses"
	"github.com/freakjazz/clean-api-go/domain/entities"
	"github.com/freakjazz/clean-api-go/domain/repositories"
)

type GetStockMovementsQuery struct {
	ProductID string
	Page      int
	PageSize  int
}

// Items is []*entities.StockMovement rather than []map[string]interface{}
// — see GetProductsResult's doc comment for why.
type GetStockMovementsResult struct {
	Items      []*entities.StockMovement
	TotalCount int
	Page       int
	PageSize   int
}

type GetStockMovementsQueryHandler struct {
	repo   repositories.IStockMovementRepository
	logger *logging.EnhancedLogger
}

func NewGetStockMovementsQueryHandler(repo repositories.IStockMovementRepository) *GetStockMovementsQueryHandler {
	return &GetStockMovementsQueryHandler{
		repo:   repo,
		logger: logging.NewEnhancedLogger("clean-api-go").WithLayer("application").WithComponent("GetStockMovementsQueryHandler").WithMethod("Handle"),
	}
}

func (h *GetStockMovementsQueryHandler) Handle(ctx context.Context, q GetStockMovementsQuery) (*GetStockMovementsResult, *bbex.ErrorResponse) {
	log := h.logger

	movements, total, err := h.repo.FindByProductID(ctx, q.ProductID, q.Page, q.PageSize)
	if err != nil {
		log.ErrorWithCode("FindByProductID failed", bberrors.InfraDBFailure.Int(), map[string]interface{}{"error": err.Error()})
		e := bbex.ErrorResponseBuilder.InternalServerError(err.Error())
		return nil, &e
	}

	return &GetStockMovementsResult{Items: movements, TotalCount: total, Page: q.Page, PageSize: q.PageSize}, nil
}

File: examples/python/clean_api_python/application/queries/get_stock_movements.py
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from backbone.infrastructure.logging import LoggerFactory

from domain.repositories.stock_movement_repository import IStockMovementRepository


@dataclass
class GetStockMovementsQuery:
    product_id: Optional[str] = None
    page: int = 1
    page_size: int = 10


@dataclass
class GetStockMovementsResult:
    items: List[Dict[str, Any]]
    total_count: int
    page: int
    page_size: int


class GetStockMovementsQueryHandler:
    def __init__(self, repo: IStockMovementRepository) -> None:
        self._repo = repo
        self._logger = LoggerFactory.create_for_layer(
            service_name="clean-api-python", layer="application", component="GetStockMovementsQueryHandler",
        )

    def handle(self, query: GetStockMovementsQuery) -> GetStockMovementsResult:
        movements, total = self._repo.find_by_product_id(query.product_id, query.page, query.page_size)
        self._logger.info(
            "Stock movements listed",
            context={"product_id": query.product_id, "total": total, "page": query.page},
        )
        return GetStockMovementsResult(
            items=[m.to_dict() for m in movements],
            total_count=total,
            page=query.page,
            page_size=query.page_size,
        )
